### DISCUSSION TITLE
Connecting to Elli Wallbox over EEBus to HEMS and sharing my experience/trial/errors in the process ...

### DISCUSSION THREAD

**Discussion by ghost** [enbility/eebus-go#71] [Show and tell]
**Goal:**
Connect Elli Wallbox over EEBus to HEMS

**Research:**
Tried to do it with OpenMuc, a Java library, as their documentation looks good: https://www.openmuc.org/eebus/jspine/javadoc/, however, there is no Download button on the page https://www.openmuc.org/eebus/jspine/. I was trying to call them and I got some emails to write, but until now, no response. I have also sent a contact form, no response **yet**.

Next try was the Go library from @DerAndereAndi . 

**Support**
Everything written here was tested with the great help of @DerAndereAndi , he was always willing to help, connect and talk over Slack (https://enbility.slack.com/).

**Connecting to Elli Wallbox Configuration page**
We are using a Linux-based controller where Elli and Reduxi are connected to the same network,.
So I tried to ping the Elli and it works.
The next step was to create an SSH tunnel to my local host machine so I can open the configuration page of the Elli Wallbox.

` ssh -i "C:\Users\user\.ssh\openssh_user" -L :443:192.168.1.200:443 root@x.x.x.x`

![image](https://github.com/enbility/eebus-go/assets/139472583/0db92637-7320-41bd-baae-ae6cb1a8da8c)

We need the Energy Manager page where we will start the pairing process with the EEBus Go Library.

**Specific Go library problems**
#1 problem: It's specific to our Linux version ... just short info. The best method for me was to install Go using https://go.dev/doc/install and remember to export PATH=$PATH:/usr/local/go/bin.

#2 problem: cert related, read this for more info - https://github.com/evcc-io/eebus/issues/1

**Running EEBus-go Go library**
It did not work for me. It's related to our specific Linux version, so @DerAndereAndi was kind enough to provide me with another option.

> You could try this app, has binaries in the releases section: https://github.com/enbility/devices-app
> Start it on the linux device and connect via the browser to the web interface. And you’ll be able to connect to the Elli Wallbox 

**Running Devices app**
Going to https://github.com/enbility/devices-app and selecting the latest version - https://github.com/enbility/devices-app/releases/tag/v0.1-20230724, choosing the right version for our Linux type. Setting correct permission and then running this.

![image](https://github.com/enbility/eebus-go/assets/139472583/a30d8187-1f9f-4f6f-ace9-2a780cadc0f4)

... next step is to access the Devices app over a web browser. If you are using SSH access to connect to the HEMS controller, you can always do an SSH tunnel, for example: 

` ssh -i "C:\Users\user\.ssh\openssh_user" -L :7050:127.0.0.1:7050 root@x.x.x.x`

**Wierd Elli software problems**
Maybe it's me doing something wrong, but I had some problems with the Energy Manager page at Elli WallBox. The service was stopping, restarting ... so the page looked like this was normal for me:
![image](https://github.com/enbility/eebus-go/assets/139472583/8b544677-70a0-45aa-bc36-4b71a4c16bff)

... this also affects pairing **Devices-app** Go Library with **Elli Wallbox**. The process started, however in the middle, the connection stopped. @DerAndereAndi  told me it's normal that the Elli Wallbox services "stops/crashes" at the first attempt, but for me, it was always starting, communicating started, stopped... and looping again.

`websocket: close 1006 (abnormal closure): unexpected EOF
`
![image](https://github.com/enbility/eebus-go/assets/139472583/d2db3e22-08fe-42bf-bd17-ee65cd0fc598)

**Checking the communication data**
@DerAndereAndi told me TRACE SEND RECV and DEBUG SEND RECV commands ... are showing there is some communication going on, however, it is stopped due to the error above.

DEBUG Recv: d:_i:47859_Elli-Wallbox-****: to NodeManagement read 13 NodeManagementDestinationListData
2023-07-24 14:34:50 DEBUG Send: d:_i:47859_Elli-Wallbox-**** reply 9 13 NodeManagementDestinationListData
2023-07-24 14:34:50 DEBUG Send: d:_i:47859_Elli-Wallbox-**** result 10 13 ResultData 0

**Why do I see EEBus-go library first connecting to IP and then to hostname?**
Because connection was lost over IP, so EEBus-go library is trying to connect over hostname:
 9 failed: dial tcp: lookup wallbox-******.local. on 192.168.1.1:53: no such host

**Going from 1,2, trying to go the 3 and back**
I did never get to the third screen ( 3. View capabilities ) ... as it always went fiding, pairing, communicating some data ... and then the connection dropped.

![image](https://github.com/enbility/eebus-go/assets/139472583/416db31c-e8b1-4ad9-8a55-efc4cd9f80b6)

**Trying to ....**
Refreshing the Elli Wallbox page, starting/stopping the devices-app library ... nothing worked. Bouncing through errors on the Elli Wallbox configuration page or not being able to go further than the 2. step, pairing the device with the devices-app library.

![image](https://github.com/enbility/eebus-go/assets/139472583/968d6b62-d983-41d5-80e8-36030c18ba55)

It's weird that sometimes both pages were saying the pairing is done, but because of the connection being dropped, it was always going back and forth.

![image](https://github.com/enbility/eebus-go/assets/139472583/0070f0ff-769c-4999-8761-09cdd6cdb26a)

`2023-07-24 12:52:15 DEBUG 676c21ca6a0e****  websocket read error:  websocket: close 1006 (abnormal closure): unexpected EOF`

**Some communication going on:**
`
2023-07-24 12:53:01 DEBUG Send: d:_i:47859_Elli-Wallbox-****result 12 15 ResultData 0
2023-07-24 12:53:01 TRACE Send: 676c21ca**** {"data":[{"header":[{"protocolId":"ee1.0"}]},{"payload":{"datagram":[{"header":[{"specificationVersion":"1.1.1"},{"addressSource":[{"device":"d:_i:EnbilityNet_Devices-App-123456789"},{"entity":[0]},{"feature":0}]},{"addressDestination":[{"device":"d:_i:47859_Elli-Wallbox-****"},{"entity":[0]},{"feature":0}]},{"msgCounter":11},{"msgCounterReference":15},{"cmdClassifier":"reply"}]},{"payload":[{"cmd":[[{"nodeManagementUseCaseData":[]}]]}]}]}}]}
2023-07-24 12:53:01 TRACE Send: 676c21ca**** {"data":[{"header":[{"protocolId":"ee1.0"}]},{"payload":{"datagram":[{"header":[{"specificationVersion":"1.1.1"},{"addressSource":[{"device":"d:_i:EnbilityNet_Devices-App-123456789"},{"entity":[0]},{"feature":0}]},{"addressDestination":[{"device":"d:_i:47859_Elli-Wallbox-****"},{"entity":[0]},{"feature":0}]},{"msgCounter":12},{"msgCounterReference":15},{"cmdClassifier":"result"}]},{"payload":[{"cmd":[[{"resultData":[{"errorNumber":0}]}]]}]}]}}]}
2023-07-24 12:53:01 DEBUG 676c21ca6a**** websocket read error:  websocket: close 1006 (abnormal closure): unexpected EOF
2023-07-24 12:53:01 DEBUG delaying connection to 676c21c****9 by 5.371s to minimize double connection probability
`

**Trying build version of EEBus-go library**
@DerAndereAndi  was kind enough to build the go library. Example ...
./main 4715 "676C 21CA 6A0E ****" cert.crt cert.key

*cert note:* running devices-app already generated cert.cert and cert.key, so I reused them running EEBus-go library.

**... this did not work due to SKI.
In this case copy paste the SKI from the console and use that SKI format (without spaces, lower case) running the library.**
./main 4715 "676c21ca6a0e****" cert.crt cert.key

... but the same problem appeared. Connecting, pairing, communication and then the communication drops.

**Checking the firmware version**
Checking at the https://www.elli.eco/en/changelog/wallbox website, there is a new firmware being released tomorrow, however, it does not say it affects any kind of EEBus-related problems. However, it would be nice to try.

**Any other solution***
@DerAndereAndi  said one option is to power out/power in (restart) the Elli Wallbox and try again. This is on my TODO list before giving up :)

... thank you @DerAndereAndi for going the extra mile and helping me understand how this works.

This research was paid by [Reduxi - Energy management System](https://reduxi.eu/en/)

**Comment by ghost** [comment]
**Some extra information.
Example of connection being dropped using device-app:**
![image](https://github.com/enbility/eebus-go/assets/139472583/f3958ce8-7d28-4a87-9ccd-da10303690e7)

**There was a small error with devices-app, so please check if you have the latest version:**
![image](https://github.com/enbility/eebus-go/assets/139472583/54b0e4a3-5fd2-454e-94a7-2030db90a072)

**Elli Wallbox recognizing the Devices-app as the possibility to pair:**
![image](https://github.com/enbility/eebus-go/assets/139472583/c7425c69-260a-44a2-8d9b-95d1389f3dcc)

**Checking with what SSL does Elli Wallbox responded:**
openssl s_client  -connect 192.168.1.200:4712

depth=0 CN = EEBUS, O = EVBox Intelligence, C = NL 
verify error:num=18:self signed certificate 
verify return:1 
depth=0 CN = EEBUS, O = EVBox Intelligence, C = NL 

**Software to try out the EEBus-go library**
Check https://github.com/evcc-io/evcc which included EEBus-go library for connecting the Elli Wallbox (among others).

**Comment by ghost** [comment]
While checking the terminal using EEBus-go library, I noticed "failed:  tls: failed to parse certificate from server: x509: invalid basic constraints". From reading https://github.com/evcc-io/eebus/issues/1, and then https://github.com/golang/go/issues/11091 and looking at the proposed patch https://github.com/golang/go/issues/11091, which was declined, I went to https://github.com/enbility/eebus-go/discussions/70 and from there to EVCC MakeFIle: https://github.com/evcc-io/evcc/blob/master/Makefile#L134-L144 ... and running following commands on Linux arm:

```
#  Change /storage/rw to the path where you uploaded/download asn1.diff patch
export GOROOT=/usr/local/go
patch-asn1-sudo::
	# echo $GOROOT
	cat $GOROOT/src/vendor/golang.org/x/crypto/cryptobyte/asn1.go | grep -C 1 "out = true"
	sudo patch -N -t -d $GOROOT/src/vendor/golang.org/x/crypto/cryptobyte -i /storage/rw/asn1.diff
	cat $GOROOT/src/vendor/golang.org/x/crypto/cryptobyte/asn1.go | grep -C 1 "out = true"
```

Source:
 - accessing $GOROOT (https://askubuntu.com/questions/917149/how-to-get-the-value-of-an-exported-variable-in-the-console)
 
 **... after this patch I still get:**
"failed to parse certificate from server: x509: invalid basic constraints".

**... although Elli Wallbox says it's connected.**
![image](https://github.com/enbility/eebus-go/assets/139472583/eddb2bd0-71ff-4571-9f4a-cef3ced2448d)

**Reply by ghost** [reply]
I forgot, I am running ./main, which was compiled on Windows. So I was patching Linux, but I am using binaries from Windows go version which was not patched. So, let's patch Windows go.

```
1. Install git bash
2. run git bash as administrator
3. run the command patch -N -t -d "C:\Program Files\Go\src\vendor\golang.org\x\crypto\cryptobyte" -i "C:\Users\user\asn1.diff"
```

... and then compiling EEBus-go with patched go:
```
1. run CMD as administrator
2. enter:> set GOOS=linux
3. enter:> set GOARCH=arm
4. enter:> go build cmd/hems/main.go
```

**... upload binaries to Linux and ... it works:**
2023-07-25 08:27:39 DEBUG 676c21ca6a0eccf7dc3cf282d255d1e75f2c7769 websocket read error:  websocket: close 1006 (abnormal closure): unexpected EOF

... so the next step is to solve this websocket read error.

**Reply by DerAndereAndi** [reply]
You can't solve this websocket read error. That is a bug in the Elli Software where the EEBUS daemon crashes and thus the websocket connection gets closed.

**Comment by ghost** [comment]
This comment was saying https://github.com/gorilla/websocket/issues/321, that increasing HandshakeTimeout fixed the problem. However, it does not work on my end after increasing value to 45 seconds.

Compiling everything on Windows, and uploading to Linux takes some time. Let's see how to compile, run and connect from Windows to Elli Wallbox trough SSH tunnel.

**1. Set back the GOARCH and GOOS**
set GOARCH=amd64
set GOOS=windows

**2. Running >go run cmd/hems/main.go 4715 --- which will generate cert files**
![image](https://github.com/enbility/eebus-go/assets/139472583/b67e703e-c54e-4485-953c-8057eb831976)

-----BEGIN EC PRIVATE KEY-------> create cert.key file and copy paste the content there (watch for spaces before/after)
 -----BEGIN CERTIFICATE----- --> create cert.crt file and copy paste the content there (watch for spaces before/after)

**3. Making mDNS available trough SSH tunnel?**
Looks like not -> https://serverfault.com/questions/121032/forward-mdns-from-one-subnet-to-another.

... trying to use https://github.com/miselin/mDNS-Unicast-Lookup with https://github.com/realk1ko/mdns-proxy/blob/main/build.sh, installing pip (https://www.tecmint.com/install-pip-in-linux/) ... but ended up with no device left and problems due to that :)

... so I have to come up with another solution :)

**Reply by DerAndereAndi** [reply]
I am sorry, but you are on the wrong track. The disconnection is because the remote device (Elli) wallbox closes the connection! And that is because the EEBUS service on it crahes.

**Reply by ghost** [reply]
You are right. However, I can not change how Elli works.
I would like to have more understanding (knowledge) and implement some of the possible solutions (trial/error) to see if anything can help Elli behave better :)

**#1 Data Race - https://go.dev/doc/articles/race_detector**
**go: -race requires cgo; enable cgo by setting CGO_ENABLED=1**
go env -w CGO_ENABLED=1

**# runtime/cgo
cgo: C compiler "gcc" not found: exec: "gcc": executable file not found in %PATH%**
(https://www.computerhope.com/issues/ch000549.htm#windows10, https://stackoverflow.com/questions/714877/setting-windows-powershell-environment-variables)
powershell,  $env:Path += ';C:\TDM-GCC-64\bin\'

... open a new CMD (as administrator) so the path variable is refreshed.

**go build -race cmd/hems/main.go**
-race is not supported on linux/arm

**Reply by ghost** [reply]
**@DerAndereAndi said one option is to power out/power in (restart) the Elli Wallbox and try again. This is on my TODO list before giving up :)**
So I got the Elli Wallbox restarted ... and now the error is: failed:  dial tcp 192.168.1.200:4712: i/o timeout ... using the binaries you sent over slack. The same result using my binaries. Weird :)
![image](https://github.com/enbility/eebus-go/assets/139472583/bf0e5105-d811-4b2b-8935-7ed8cfb08fc6)
